sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java

```diff
@@ -63,8 +83,73 @@ public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, P
         postRegister();
     }
 
+    private void processOffers() {
+        offerExecutor.execute(() -> {
+            while (true) {
+                // This is a blocking call which pulls as many elements from the offer queue as possible.
+                List<Protos.Offer> offers = offerQueue.takeAll();
+                processOfferSet(offers);
+                offersInProgress.removeAll(
+                        offers.stream()
+                                .map(offer -> offer.getId())
+                                .collect(Collectors.toList()));
+            }
+        });
+    }
+
+    protected abstract void processOfferSet(List<Protos.Offer> offers);
+
+    @Override
+    public void resourceOffers(SchedulerDriver driver, List<Protos.Offer> offers) {
+        if (!apiServerReady()) {
+            LOGGER.info("Waiting for API Server to start ...");
+            OfferUtils.declineOffers(driver, offers);
+            return;
+        }
+
+        // Task Reconciliation:
+        // Task Reconciliation must complete before any Tasks may be launched.  It ensures that a Scheduler and
+        // Mesos have agreed upon the state of all Tasks of interest to the scheduler.
+        // http://mesos.apache.org/documentation/latest/reconciliation/
+        reconciler.reconcile(driver);
+        if (!reconciler.isReconciled()) {
+            LOGGER.info("Waiting for task reconciliation to complete...");
+            OfferUtils.declineOffers(driver, offers);
+            return;
+        }
+
+        offersInProgress.addAll(
+                offers.stream()
+                        .map(offer -> offer.getId())
+                        .collect(Collectors.toList()));
+
+        for (Protos.Offer offer : offers) {
+            boolean queued = offerQueue.offer(offer);
+            if (!queued) {
+                LOGGER.warn("Failed to enqueue offer: '{}'", offer.getId().getValue());
+                OfferUtils.declineOffers(driver, Arrays.asList(offer));
+            }
+        }
+    }
+
     protected abstract void initialize(SchedulerDriver driver) throws InterruptedException;
 
+    protected abstract boolean apiServerReady();
+
+    /**
+     * All offers must have been presented to resourceOffers() before calling this.
+     * @throws InterruptedException
+     */
+    @VisibleForTesting
+    public void awaitOffersProcessed() throws InterruptedException {
+        while (!offersInProgress.isEmpty()) {
```

Comment:
I don't want to put this more complicated logic which is really only needed for tests into the heart of the the offer processing loop.  Added locks, but keeping polling on `awaitOffersProcessed`.